artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueueConsumersImpl.java

```diff
@@ -0,0 +1,124 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.activemq.artemis.core.server.impl;
+
+import org.apache.activemq.artemis.core.PriorityAware;
+import org.apache.activemq.artemis.utils.collections.PriorityCollection;
+import org.apache.activemq.artemis.utils.collections.ResettableIterator;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.Set;
+import java.util.Spliterator;
+import java.util.concurrent.CopyOnWriteArraySet;
+import java.util.concurrent.atomic.AtomicReferenceFieldUpdater;
+import java.util.function.Consumer;
+
+/**
+ * This class's purpose is to hold the consumers.
+ *
+ * CopyOnWriteArraySet is used as the underlying collection to the PriorityCollection, as it is concurrent safe,
+ * but also lock less for a read path, which is our HOT path.
+ * Also it was the underlying collection previously used in QueueImpl, before we abstracted it out to support priority consumers.
+ *
+ * There can only be one resettable iterable view,
+ * A new iterable view is created on modification, this is to keep the read HOT path performent, BUT
+ * the iterable view changes only after reset so changes in the underlying collection are only seen after a reset,
+ *
+ * All other iterators created by iterators() method are not reset-able and are created on delegating iterator().
+ *
+ * @param <T> The type this class may hold, this is generic as can be anything that extends PriorityAware,
+ *         but intent is this is the QueueImpl:ConsumerHolder.
+ */
+public class QueueConsumersImpl<T extends PriorityAware> implements QueueConsumers<T> {
+
+   private final PriorityCollection<T> consumers = new PriorityCollection<>(CopyOnWriteArraySet::new);
+   private final Collection<T> unmodifiableConsumers = Collections.unmodifiableCollection(consumers);
+   private final AtomicReferenceFieldUpdater<QueueConsumersImpl, ResettableIterator> changedIteratorFieldUpdater =  AtomicReferenceFieldUpdater.newUpdater(QueueConsumersImpl.class, ResettableIterator.class, "changedIterator");
+   private volatile ResettableIterator<T> changedIterator;
+   private ResettableIterator<T> currentIterator = consumers.resettableIterator();
+
+   @Override
+   public Set<Integer> getPriorites() {
+      return consumers.getPriorites();
+   }
+
+   @Override
+   public boolean hasNext() {
+      return currentIterator.hasNext();
+   }
+
+   @Override
+   public T next() {
+      return currentIterator.next();
+   }
+
+   @Override
+   public QueueConsumers<T> reset() {
+      if (changedIterator != null) {
+         currentIterator = changedIterator;
+         changedIteratorFieldUpdater.compareAndSet(this, changedIterator, null);
```

Comment:
What happen if the cas will fail? we ends with a reset() that is not nulling `changedIterator`.
Given that we are just clearing the `changedIterator` to `null` a `lazySet(this, null) is enough.

Reply:
If it fails it means whilst we were reseting and changing reference to the new iterator. a consumer was added or removed and another new iterator was created (as this can occur whilsy we are iterating or resetting.) And means simply whilst this reset updated its iterator, on next reset it needs to switch over its again. We wouldnt want to lazy set to null

Reply:
currentIterator is ONLY updated by reset, which should ONLY be called by the same threads operating on the ResettableIterator interface.

When add or remove of consumer occurs a new iterator is parked into the volatile changedIterator (using an atomic field updater), so the next reset can pick it up.

Reply:
Thats fair comment. Ill add this.

Reply:
ill mark comment resolved, if you think it needs further docs or could be further abstracted to be more clearer let me know.